### ISSUE TITLE
Model does not select appropiate Features

### ISSUE THREAD

**Issue by AlbertoCastelo** [dreamquark-ai/tabnet#188]
Hello guys. 

Awesome job with this library :rocket:. I have an issue with it but is neither a bug or a feature request. It is more about the model itself.

I did a spike using TabNet to better understand if this model could be useful to provide explainability on timeseries data. All my code is in this repository: [Spike](https://github.com/AlbertoCastelo/tabnet-timeseries-spike)

I was expecting that the model could learn the actual important features but it couldn't. I run two scenarios. In the 1st one I select the important features. While in the second one, I provide many more features. I would expect the algorithm to be able to identify which features are more important but it couldn't. 

Do you know if what could be the reason behind it or what would you try to make it work?

Thanks. 

PD: this notebook contains the experiment I run https://github.com/AlbertoCastelo/tabnet-timeseries-spike/blob/master/notebooks/tabnet-does-not-learn-interesting-features.ipynb

**Comment by Optimox** [comment]
Hello @AlbertoCastelo,

Interesting experiment. Thanks for sharing!

## About the experiment
First let me wrap up what I understood from your experiment (please correct me if I'm wrong):
- you have a time serie problem that you are treating with auto regressive features that you are creating from past time points
- you already know that feature 1 and 12 have much more predictive powers than others
- training with 2 best features --> validation MSE goes down to 500
- training with 12 features --> validation MSE goes down to 4500
- moreover you would expect to see best features from the second model to be feature1 and feature12?

Did I understand the experiment right?

## My comments

- I agree that it's indeed a bit disappointing that the default parameters won't find the best features and get at least a similar score from the first experiment. Even if low predictive features can add noise, you would definitely expect to be able to get a similar score.
- I see two problems : being able to get similar scores and having the good feature importance. (I think it's not reasonable to ask for a good feature importance with a model that performs really poorly)

## How to get similar results on both experiment

I don't know if it's going to work but I think you should try some parameter tuning to get similar scores.
Here is what I would try:
- `n_steps=1` : when you have multiple steps, the masks depend from previous steps and depending on `gamma` they could be forced to be exclusive, so your steps will eventually select poor features as their prior don't allow them to pick always the two best. Trying with only one step could avoid overfitting.
- `lambda_sparse` : since your model seem to be looking at more features than necessary you could try to set a bigger sparsity penalty but increasing `lambda_sparse`. The range of `lambda_sparse` is hard to define because it depends on your MSE scores, you could try a few different values like 1e-3 1e-1, 1, 10, 100 and see if it helps with the validation score
- `n_d` and `n_a` : you could try to reduce the model capacity by setting them to 4 or even 2... (not a big fan of this but it may help)
- all parameters could be important, but the main goal would be to try to match the first experiment score, I hope you'll get there.

## Feature Importance

Hopefully with similar scores you'll get a feature importance that looks like what you expect.
It's also interesting to have a look at the explain matrix to see if the model is using different features for different predictions.


I hope this will help you.
I'm very interested to know about how it goes, please let us know!

**Comment by AlbertoCastelo** [comment]
Hey. Thanks for the thorough reply.

You got the experiment right. If followed on your advice and run a hyperparameter search on those parameters. The validation score obviously improves for both experiments in the best cases however, I was not able to get close to the 2-features experiment. 

Reviewing the learning curves (that you can see in this [notebook](https://github.com/AlbertoCastelo/tabnet-timeseries-spike/blob/master/notebooks/trying-seb-ideas.ipynb)), we can see that the learning process is very unstable regarding the validation set while the training-loss is much more smooth. That looks weird to me. 

![Screenshot from 2020-09-20 10-36-09](https://user-images.githubusercontent.com/25732484/93707263-2456f480-fb2d-11ea-856e-2c36c09da8d9.png)

I'm thinking on starting to review the learning process and maybe try different learning algos, rates, scheduling, etc. What do you think?

Thanks.

**Comment by Optimox** [comment]
So what is your best score in validation/test set set with the hyper search (for second experiment)?

The thing is usually lambda_sparse should be small compared to the loss, but since you are doing a regression task the ranges are not clear. So maybe every experiment with lambda sparse >1 does not make sense (and I can't tell which is which in your notebook).

One other thing is that you are plotting the losses including the sparsity loss, which could be the reason why things are fluctuating so much. Maybe you could try to plot `self.metrics_valid` which would be only what people are usually interested in : your MSE score. (plotting `self.losses_valid - self.metrics_valid` would give you the value of the sparsity loss, so you can decide what would be a reasonnable lambda_sparse for this experiment).

If you still see weird fluctuations (the one you shared looks ok to me by the way, TabNet can take some time before converging on validation, and then the explosion is weird but why not but some others look terrible I agree) when looking at `metrics_valid` you could indeed lower the learning rate (either from the start or by using a scheduler).

Thanks for sharing, let me know how it goes.

**Comment by Optimox** [comment]
@AlbertoCastelo do you have some news on this?

**Comment by Optimox** [comment]
closing this feel free to reopen if you have new information to share
